### ISSUE TITLE
the moving mean and moving variance of batch_norm seems not being saved

### ISSUE THREAD

**Issue by dixonhsiao** [zhengyang-wang/Unet_3D#5]
As title, it seems that in test/predict mode the output of batch_norm is very weird, I believe it is due to the moving mean and variances are not saved.

so in network.py use the following code:

def configure_networks(self):
        self.build_network()
        update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
        with tf.control_dependencies(update_ops):
            optimizer = tf.train.AdamOptimizer(self.conf.learning_rate)
            self.train_op = optimizer.minimize(self.loss_op, name='train_op')
        tf.set_random_seed(self.conf.random_seed)
        self.sess.run(tf.global_variables_initializer())
        self.saver = tf.train.Saver()
        self.writer = tf.summary.FileWriter(self.conf.logdir, self.sess.graph)

insted of using :
trainable_vars = tf.trainable_variables()
self.saver = tf.train.Saver(var_list=trainable_vars, max_to_keep=0)
use :
self.saver = tf.train.Saver()
so that the moving mean and variace are saved.

also, for batch_norm function, use a train_flag and put it into a bool placeholder(self.phase), remember to modify all calls to conv, deconv, pixel_dcl, ipixel_cl, ipixel_dcl functions:
in network.py:
def build_network(self):
   ......
   self.phase = tf.placeholder(tf.bool)  #define a bool placeholder
...
conv1 = ops.conv(inputs, out_num, self.conv_size,
                         name+'/conv1', self.conf.data_type, train_flag=self.phase)
...

in ops.py:
def conv(inputs, out_num, kernel_size, scope, data_type='2D', norm=True, train_flag=False):
   ......
    if norm:
        return tf.contrib.layers.batch_norm(
            outs, decay=0.9, epsilon=1e-5, activation_fn=tf.nn.relu,
            updates_collections=tf.GraphKeys.UPDATE_OPS, scope=scope+'/batch_norm', is_training=train_flag)
   ......

when train:
feed_dict = {self.inputs: inputs,
                   self.annotations: annotations,
                   self.phase: True}

when test/predict:
feed_dict = {self.inputs: inputs,
                   self.annotations: annotations,
                   self.phase: False}

**Comment by dixonhsiao** [comment]
![image](https://user-images.githubusercontent.com/33412775/44956298-5b312200-aef4-11e8-987c-b40c23e1f68e.png)

**Comment by dixonhsiao** [comment]
sorry I don't know how to do indentation in the github, so it's a little bit hard to read.

**Comment by zhengyang-wang** [comment]
Thank you. It is a known issue to me. Check [here](https://github.com/zhengyang-wang/Unet_3D/blob/master/model/README.md#note-important).

Your modification about is_training is absolutely correct as I suggested in the note, and it is a crucial step to using operations that operate differently in training and testing phases, like batch_norm and dropout. About the moving_mean and moving_variance not saved, I will check the problem and modify the code accordingly.

Actually, I implemented a correct version of Unet-3D in this [project](https://github.com/zhengyang-wang/3D-Unet--Tensorflow). If you are only interested in Unet-3D, you could use this version.

**Comment by zhengyang-wang** [comment]
As I set updates_collections=None in batch_norm, there is no need to add

> update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
> with tf.control_dependencies(update_ops):

You can use [inspect_checkpoint](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/inspect_checkpoint.py) to check whether moving_mean and moving_variance have been saved.
As a result, I think if you follow the note, the batch_norm will work correctly.
